gdnative-core/src/core_types/vector3.rs

```diff
@@ -10,64 +20,176 @@ pub enum Axis {
 
 /// Helper methods for `Vector3`.
 ///
-/// Trait used to provide additional methods that are equivalent to Godot's methods.
 /// See the official [`Godot documentation`](https://docs.godotengine.org/en/3.1/classes/class_vector3.html).
-pub trait Vector3Godot {
-    /// Internal API for converting to `sys` representation. Makes it possible to remove
-    /// `transmute`s elsewhere.
+impl Vector3 {
+    pub const ZERO: Self = Vector3::new(0.0, 0.0, 0.0);
+    pub const ONE: Self = Vector3::new(1.0, 1.0, 1.0);
+
+    #[inline]
+    pub const fn new(x: Real, y: Real, z: Real) -> Self {
+        Self { x, y, z }
+    }
 
     /// Returns the normalized vector pointing from this vector to `other`.
-    fn direction_to(self, other: Vector3) -> Vector3;
+    #[inline]
+    pub fn direction_to(self, other: Vector3) -> Vector3 {
+        (other.glam() - self.glam()).normalize().into()
+    }
+
     /// Returns the distance to `other`.
-    fn distance_to(self, other: Vector3) -> f32;
+    #[inline]
+    pub fn distance_to(self, other: Vector3) -> f32 {
+        other.glam().distance(self.glam())
+    }
+
     /// Returns the squared distance to `other`.
     ///
     /// This method runs faster than distance_to, so prefer it if you need to compare vectors or
     /// need the squared distance for some formula.
-    fn distance_squared_to(self, other: Vector3) -> f32;
+    #[inline]
+    pub fn distance_squared_to(self, other: Vector3) -> f32 {
+        other.glam().distance_squared(self.glam())
+    }
+
+    #[inline]
+    pub fn length(self) -> Real {
+        self.glam().length()
+    }
+
+    #[inline]
+    pub fn normalize(self) -> Self {
+        self.glam().normalize().into()
+    }
+
+    #[inline]
+    pub fn dot(self, other: Self) -> Real {
+        self.glam().dot(other.glam())
+    }
+
+    #[inline]
+    pub fn cross(self, other: Self) -> Self {
+        self.glam().cross(other.glam()).into()
+    }
+
+    // TODO This should be is_equal_approx (hence why it's pub(crate) for now)
+    // Implementation shamelessly stolen from the euclid crate
+    #[inline]
+    pub(crate) fn approx_eq(self, other: &Self) -> bool {
+        let approx_epsilon = 1e-6;
+        (self.x - other.x).abs() < approx_epsilon
+            && (self.y - other.y).abs() < approx_epsilon
+            && (self.z - other.z).abs() < approx_epsilon
+    }
+
+    #[inline]
+    pub(crate) fn to_array(self) -> [Real; 3] {
+        [self.x, self.y, self.z]
+    }
```

Comment:
It should be possible to express this as `AsRef<[f32; 3]>` instead, since the type is `repr(C)`.

Reply:
The `euclid` version does return an array though, and replacing the return type with `AsRef<[f32; 3]>` gives me:

```
error[E0308]: mismatched types
   --> gdnative-core/src/core_types/geom/basis.rs:331:17
    |
331 |                 matrix.elements[0].to_array(),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected array `[f32; 3]`, found trait object `dyn AsRef`
    |
    = note:     expected array `[f32; 3]`
            found trait object `(dyn AsRef<[f32; 3]> + 'static)`
```

Reply:
What I meant is that you could add that as a separate impl:

```rust
impl Vector3 {
    fn to_array(&self) -> [f32; 3] {
        *self.as_ref()
    }
}

impl AsRef<[f32; 3]> for Vector3 {
    fn as_ref(&self) -> &[f32; 3] {
        // SAFETY: Vector3 is repr(C)
        unsafe { &*(self as *const Vector3 as *const [f32; 3]) }
    }
}
```